csharp/src/Ice/OutgoingFrame.cs

```diff
@@ -133,28 +143,31 @@ public CompressionResult CompressPayload()
             }
             else
             {
-                IList<ArraySegment<byte>> encapsulation = Data.Slice(_encapsulationStart, _payloadEnd);
+                IList<ArraySegment<byte>> payload = Payload;
+                int encapsulationOffset = this is OutgoingResponseFrame ? 1 : 0;
+
+                // The encapsulation always starts in the first segment of the payload (at position 0 or 1).
+                Debug.Assert(encapsulationOffset < payload[0].Count);
 
-                int sizeLength = Protocol == Protocol.Ice2 ? 1 << (encapsulation[0][0] & 0x03) : 4;
-                byte compressionStatus = encapsulation[0].Count > sizeLength + 2 ?
-                    encapsulation[0][sizeLength + 2] : encapsulation[1][sizeLength + 2 - encapsulation[0].Count];
+                int sizeLength = Protocol == Protocol.Ice2 ? payload[0][encapsulationOffset].ReadSizeLength20() : 4;
```

Comment:
I can't comment on it since it wasn't changed by this PR but somewhat related to this line... a few lines below we call `OutputStream.WriteFixedLengthSize20` without checking the protocol version. This doesn't sound correct as if the protocol is Ice1 the encapsulation size length should be 4?

Reply:
This is actually correct.

This payload compression is about compressing an encapsulation encoded in the 2.0 encoding. This encapsulation can be carried by an ice1 frame, in which case the encapsulation header itself is encoded with 1.1, with a size length of 4; more commonly, the frame is an ice2 frame and the encapsulation header is 2.0 and the encapsulation size is a variable-length.

The `OutputStream.WriteFixedLengthSize20` a few lines below is about writing the size of the uncompressed data _inside_ this 2.0 encoded encapsulation. This size is always 2.0-encoded, like every other bytes in these 2.0-encoded encapsulations.